leia-refresher/src/main/java/com/grookage/leia/provider/TimeBasedDataProvider.java

```
@@ -90,23 +96,34 @@ public Updater() {
         }
 
         public void run() {
-            try {
-                T data = configSupplier.get();
-                if (data != null) {
-                    if (Boolean.TRUE.equals(shouldUpdate.apply(dataReference.get(), data))) {
-                        dataReference.set(data);
-                        lastUpdatedTimestamp.set(System.currentTimeMillis());
-                        log.info("[LeiaRefresher.update] Successfully Updated data reference for {}..", supplierName);
-                    } else {
-                        log.info("[LeiaRefresher.update] Failed because shouldUpdate returned false, supplierName: {}", supplierName);
-                    }
+            refreshData();
+        }
+    }
+
+    private boolean refreshData() {
```

Don't do it like this. Use an atomicReference AtomicReference<Boolean> let's say 

private final AtomicReference<Boolean> initialRefresh = new AtomicReference<>(false). Set it to true after the initialRefresh and throw an exception if it is false. Don't do the return type ..

---

Added a failureHandler instead of returning boolean